### DISCUSSION TITLE
Welcome to incus-compose Discussions!

### DISCUSSION THREAD

**Discussion by bketelsen** [bketelsen/incus-compose#16] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by vmsman** [comment]
I run the "ScottiBYTE" YouTube channel and I think that many folks might be interested in this project.  I can't find any explanation or documentation here.  Is this a front end to Incus OCI containers?

**Reply by bketelsen** [reply]
Hey thanks for all the great tutorials & content, I've enjoyed them!

There's no documentation yet because it's not quite ready for public usage.  It is an implementation of the docker-compose spec that uses incus as the container engine. That means it's not at all limited to OCI containers, and can use normal incus/lxc images. That is my personal primary use-case for it. It works with OCI because incus works with OCI.

The project works today for a subset of the most common compose features. There are some things to work out still, like health checks - that feature is built into the docker/podman engines so docker-compose just piggy-backs on those check definitions. Incus has no similar feature. We could implement them from scratch, but that would require a daemon running that checks the health of containers/instances all the time. Since incus-compose is a cli tool, this would require us writing and users installing some service that does these health checks. it's not out of the realm of possibility, but I'd like to think it through before diving in there.

Docker compose also supports scaling, replicas, load balancers with the `deploy:` fields. This is another area that would require a lot of thought to map into the incus world.

Networking is a completely different beast in incus. Today incus-compose will create a new bridge with default settings for your compose project, or it will connect to an existing incus network. Anything more complex than that doesn't work. Connecting an instance to multiple networks won't work out of the box for most scenarios, either. Docker wires that up behind the scenes.

With all of those caveats (and more... see GH issues), it does work today as long as you don't stray from the core features. All of the samples in the repo work, and there are some for incus images and oci images to compare. Because health checks don't exist, you might see failures that require instance restarts.  The ghost sample is an example. The ghost app will fail and stop because it can't connect to the db. You can wait a second or two and start it and everything is fine.
